### PR TITLE
Use a specific lock file to avoid temp directory contention

### PR DESCRIPTION
**Pull request by sbc100**
Previously in debug mode we could try to avoid temp file contension by
taking and holding the cache lock for the duration of the process.

This change uses a specific lock file within the temp directory which
avoids needing a writable cache directory. Its also more logical.

Partial fix for #13369

### CODE REVIEW

**Comment by sbc100** [comment]
This change is needed in order to unblock the emsdk bump (current blocked on bazel and #13369)

File: tools/shared.py
Comment (kripken) [inline tools/shared.py:400]:
```suggestion
      # Since the cannonical temp directory is, by definition, the same
```

File: tools/shared.py
Comment (kripken) [inline tools/shared.py:402]:
```suggestion
      # process lock to prevent more than one process from writing to it.
```

File: tools/shared.py
Comment (kripken) [inline tools/shared.py:408]: it is a little unfortunate to mention `EM_EXCLUSIVE_CACHE_ACCESS` here as it would be the first time outside of `cache.py`. Perhaps we could call a helper function in that file?

**Review by kripken — APPROVED** [review]
lgtm % comments

If it's urgent to land this we can defer the comments for later.

File: tools/shared.py
Comment (sbc100) [reply tools/shared.py:408]: It is a little sad that need this exception here.. but I'm OK with hardcoding it for now.    Hopefully in the long run we can completely remove EM_EXCLUSIVE_CACHE_ACCESS.

**Pull request merged**
